### PR TITLE
Picard iteration refactoring

### LINKED ISSUE #9038: Steady and eigen executioners do not have Picard iteration
### Description of the enhancement or error report
Should we add a simple outer Picard loop?

### Rationale for the enhancement or information for reproducing the error
This is required by some of the steady-state multiphysics simulations. Also if this is in place, codes in Rattlesnake for doing nonlinear diffusion acceleration can be cleaned up.

### Identified impact
(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)
New capability so no impact on the current capabilities.

### PR DESCRIPTION
**Pull request by YaqiWang**
Close #9038.

As discussed, I made compromise for possible super loop Executioner. Instead of put the Picard loop in `FEProblemBase::solve`, I created an object within `Executioner` as `Executioner::PicardSolve` and let it to drive the Picard iteration.

This change passes all framework and module tests on my mac. I expect its impact to application is small. Rattlesnake will be broken because a virtual function, override by one of Rattlesnake executioner, in `Transient` is removed.

This is the bulk part of the refactoring. Please review before I go further.

### CODE REVIEW

File: framework/include/executioners/Executioner.h [hunk 1]
```diff
@@ -109,6 +109,141 @@ class Executioner : public MooseObject,
    */
   virtual bool lastSolveConverged();
 
+  class PicardSolve : public MooseObject, public PerfGraphInterface
```
Comment (friedmud) [inline framework/include/executioners/Executioner.h]: This needs to be in its own file - probably in the executioners directory

**Review by friedmud — CHANGES_REQUESTED** [review]
Just trying to give you a quick turnaround comment...

This definitely looks pretty good... can you put the PicardSolve in its own file then I'll do more reviewing?

File: framework/include/executioners/Executioner.h (hunk 1 above)
Comment (YaqiWang) [reply framework/include/executioners/Executioner.h]: Sure.

**Comment by YaqiWang** [comment]
This PR now contains all changes I make in the old one. It is ready for review. One small thing I noticed is that there could be a unnecessary postprocessor evaluation in `IterationAdaptiveDT::converged`. I remove it in 3467268. I did not check if this removing is possible after this refactoring but did check that all tests in framework, modules and Bison pass. @bwspenc might be interested in and possibly can make a comment on this commit.

**Comment by YaqiWang** [comment]
Added one more commit to allow derived executioner to augment Picard convergence check. For doing it, I made only executioner can construct `PicardSolve`. I believe this is a good move. The augmented check is needed to patch Rattlesnake later.

**Comment by friedmud** [comment]
Ok - I'm probably not going to be able to get to this today, but maybe Monday.

**Comment by YaqiWang** [comment]
Thanks for letting me know.

**Comment by YaqiWang** [comment]
`postSolve` is called twice with 5b34ae9. I will push a fix shortly.

**Review by friedmud — APPROVED** [review]
👍

**Comment by friedmud** [comment]
@YaqiWang Thanks for working with us on this.  I think that what we ended up with here is _really_ good for MOOSE and provides a good direction for the future.

Thanks again!

**Comment by YaqiWang** [comment]
Thanks! I will patch up the apps. I agree this the good direction, i.e. separating the solver out from problem. I can see the potentials of ultimate flexibilities with this design.

**Comment by permcody** [comment]
@YaqiWang - You might want to hold off on putting PRs into apps for a bit. The next branch is a mess right now due to https://github.com/idaholab/moose/pull/12819. I'm waiting for a fix and will hopefully be able to merge it in tonight, or early tomorrow. If you attempt to update to either HEAD or your merged PR, you'll be broken for other reasons.

**Comment by YaqiWang** [comment]
Also @lindsayad 's PR breaks Rattlesnake as well. I need that to be fixed as well to proceed.

**Pull request merged**